### PR TITLE
Refactor the parameter type of add function

### PR DESCRIPTION
**Pull request by jeffreyhc**
Ref: https://github.com/chewing/chewing-editor/pull/149#discussion_r57522841

I refactor the add function parameter type and rename the slot add function in UserphraseModel
to addWrapper for temporary, if there is a better idea please tell me.

### CODE REVIEW

**Comment by jserv** [comment]
@jeffreyhc : 
The rename to `addWrapper` violates the naming schema, which consists of lowercase verb and the target object begining with a capital letter. Your intention was originally "add" rather than adding "Wrapper" object to somewhere.

**Comment by jeffreyhc** [comment]
@jserv :
Thanks for your advice! How about `addPhrase`?
(I think this maybe more precise but this function name is same as a function in UserphraseView.)

File: src/view/UserphraseView.cpp [hunk 1]
```diff
@@ -26,8 +26,8 @@ UserphraseView::UserphraseView(QWidget *parent)
     :QListView{parent}
     ,UserphraseDialog_{new UserphraseDialog{this}}
     ,menu_(new UserphraseViewMenu{this})
-    ,originalPhrase{new QString()}
-    ,originalBopomofo{new QString()}
+    ,originalphrase_(QString())
+    ,originalbopomofo_(QString())
```
Comment (Chocobo1) [inline src/view/UserphraseView.cpp]: redundant init, you can remove it.

File: src/view/UserphraseView.cpp [hunk 2]
```diff
@@ -71,10 +71,10 @@ void UserphraseView::addPhrase(int result)
         return;
     }
 
-    std::shared_ptr<QString> phrase{new QString(UserphraseDialog_->getPhrase())};
-    std::shared_ptr<QString> bopomofo{new QString(UserphraseDialog_->getBopomofo())};
+    QString phrase = QString(UserphraseDialog_->getPhrase());
+    QString bopomofo = QString(UserphraseDialog_->getBopomofo());
```
Comment (Chocobo1) [inline src/view/UserphraseView.cpp]: `QString bopomofo = UserphraseDialog_->getBopomofo();`

File: src/model/UserphraseModel.cpp [hunk 3]
```diff
@@ -157,9 +157,9 @@ void UserphraseModel::refresh()
     emit refreshCompleted(userphrase_.size());
 }
 
-void UserphraseModel::add(std::shared_ptr<QString> phrase, std::shared_ptr<QString> bopomofo)
+void UserphraseModel::addWrapper(QString &phrase, QString &bopomofo)
```
Comment (Chocobo1) [inline src/model/UserphraseModel.cpp]: why not just remove this method and move the private `add()` to public?
It's not like the private `add()` is doing anything special that need to hide inside.

File: src/model/UserphraseSortFilterProxyModel.cpp [hunk 4]
```diff
@@ -28,9 +28,9 @@ UserphraseSortFilterProxyModel::UserphraseSortFilterProxyModel(QObject *parent)
 {
 }
 
-void UserphraseSortFilterProxyModel::add(std::shared_ptr<QString> phrase, std::shared_ptr<QString> bopomofo)
+void UserphraseSortFilterProxyModel::add(QString &phrase, QString &bopomofo)
```
Comment (Chocobo1) [inline src/model/UserphraseSortFilterProxyModel.cpp]: add `const`:
`add(const QString &phrase, const QString &bopomofo)`

File: src/view/UserphraseView.cpp (hunk 1 above)
Comment (jeffreyhc) [inline src/view/UserphraseView.cpp]: You mean the `originalphrase_` and `originalbopomofo_` don't need to initialize here?
So I just assign `originalphrase_ = userphrase->phrase_` and `originalbopomofo_ = userphrase->bopomofo_`?

File: src/model/UserphraseModel.cpp (hunk 3 above)
Comment (jeffreyhc) [inline src/model/UserphraseModel.cpp]: @Chocobo1 :
I thought to keep the code structure and only modified the parameter type,
if you think the private `add()` can be moved to public, and I will do the change.

File: src/view/UserphraseView.cpp (hunk 1 above)
Comment (Chocobo1) [inline src/view/UserphraseView.cpp]: >You mean the originalphrase_ and originalbopomofo_ don't need to initialize here?

Yes.

File: src/model/UserphraseModel.cpp (hunk 3 above)
Comment (Chocobo1) [inline src/model/UserphraseModel.cpp]: also tag @jserv, he should have the final word.

**Comment by jserv** [comment]
I agree with @Chocobo1 that `add` method with access specifier 'public` should be straightforward.

**Comment by jeffreyhc** [comment]
@Chocobo1 @jserv 
I've fixed all you mentioned and modify my git commit message.

**Comment by jserv** [comment]
1. The description "from private to public" is quite ambiguous for people to distinguish from phrases and dictionary storage. You should mention the access specifier.
2. Please explain the statement 'because "QString" can "implicit sharing"'.

**Comment by jeffreyhc** [comment]
@jserv 
Does the "access specifier" means slot and signal?

**Comment by jserv** [comment]
Read C++ Programming specification for the terminologies. http://en.cppreference.com/w/cpp/language/access

**Comment by jeffreyhc** [comment]
@jserv 
I think I don't get your point
> The description "from private to public" is quite ambiguous for people to distinguish from phrases and dictionary storage.

What is "dictionary storage"? Why do people can't distinguish phrases and dictionary storage?
> You should mention the access specifier.

I thought "from private to public" mentioned about the access specifier?

**Comment by jserv** [comment]
@jeffreyhc Because Chewing does have "private" dictionary! It is so-called user-defined phrase in the form of "hash", which is historically a terminology existing for more than 15 years. In the opposite side, Chewing have the system dictionary, which is "public".

**Comment by jserv** [comment]
Userphrase is kept in private storage for each IM instance, and the way to change its descriptions might result in the misunderstanding about its transition to public via shared networks.

**Comment by jeffreyhc** [comment]
@jserv 
I plan to modify my commit message like this:
```
Refactor the parameter type of add function

Remove "shared_ptr" type for the QString buffer phrase and bopomofo,
because "QString" can "implicit sharing" just do the same thing of
"shared_ptr". "Implicit sharing" makes ownership of an object can be
shared through pointers and the object is copied only if and when
the content is going to be changed, i.e., copy-on-write.
Moreover, in UserphraseModel, I remove the public add() which only
calls the private add() and the private add() is the function really add
phrase to dictionary. Then modify the access specifier of the add()
from private to public.
```
How do you think about the message?

**Comment by jserv** [comment]
@jeffreyhc : Do after review! Update frequently, and everyone can provides the feedback often.

**Comment by jeffreyhc** [comment]
@jserv 
OK, I update my commit.

**Comment by jserv** [comment]
Fix the syntax error in commit messages:
> "QString" can "implicit sharing" just do the same thing of "shared_ptr"

**Comment by jeffreyhc** [comment]
How about this version?

**Comment by jserv** [comment]
Don't use quotation marks to specify a common term or phrase such as "QString" in Git commit messages. Also, you don't have to inform others of the time when the revised patch was pushed since CI does this well.

**Pull request merged**